cmd/shortener/main.go

```diff
@@ -49,24 +54,89 @@ func main() {
 
 	config.Parse()
 
-	if err := run(); err != nil {
+	if err := run(ctx); err != nil {
 		panic("unexpected error: " + err.Error())
 	}
 }
 
-func run() error {
-	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-	defer cancel()
-
+func run(ctx context.Context) error {
 	storage, err := newStore(ctx)
 	if err != nil {
 		return fmt.Errorf("cannot create storage: %w", err)
 	}
 	defer storage.Close()
+	removeChan := make(chan models.BatchRemoveRequest)
+	semaphore := make(chan struct{}, 5)
+	instance := app.NewInstance(config.BaseURL, storage, removeChan)
+	go func() {
+		for removeRequest := range removeChan {
+			semaphore <- struct{}{}
+			go func(req models.BatchRemoveRequest) {
+				err := storage.DeleteUsers(ctx, req.UID, req.Ids...)
+				if err != nil {
+					logrus.Errorf("Couldn't delete urls for user %s", req.UID.String())
+				}
+				<-semaphore
+			}(removeRequest)
+		}
+	}()
+	srv := &http.Server{
+		Addr:    config.RunPort,
+		Handler: newRouter(instance),
+	}
+
+	if config.UseTLS {
+		err := config.MakeKeys(config.CertFile, config.KeyFile)
+		if err != nil {
+			return err
+		}
+		srv.TLSConfig = &tls.Config{
+			MinVersion: tls.VersionTLS13,
+		}
+		go func() {
+			if err := srv.ListenAndServeTLS(config.CertFile, config.KeyFile); err != nil && err != http.ErrServerClosed {
+				logrus.Fatalf("listen and serve: %v", err)
+			}
+		}()
+	} else {
+		go func() {
+			if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+				logrus.Fatalf("listen and serve: %v", err)
+			}
+		}()
+	}
 
-	instance := app.NewInstance(config.BaseURL, storage)
+	logrus.Infof("listening on %s", config.RunPort)
 
-	return http.ListenAndServe(config.RunPort, newRouter(instance))
+	<-ctx.Done()
+
+	logrus.Info("shutting down server gracefully")
+	idleConnectionsClosed := make(chan struct{}, 1)
+
+	shutdownCtx, cancel := context.WithTimeout(context.Background(), config.ShutdownTimeout)
+	defer cancel()
+
+	go func() {
+		if err := srv.Shutdown(shutdownCtx); err != nil {
+			logrus.Errorf("shutdown: %v", err)
+		}
+		close(idleConnectionsClosed)
+	}()
+
+	for {
+		select {
+		case <-shutdownCtx.Done():
+			return fmt.Errorf("server shutdown: %w", ctx.Err())
```

Comment:
Скорее всего, ошибка будет выглядеть одинаково, так как сюда мы провалимся только после прерывания обоих контекстов, но, наверное, правильнее будет здесь использовать
```suggestion
			return fmt.Errorf("server shutdown: %w", shutdownCtx.Err())
```
даже не "наверное", а "точно". Потому что ctx будет говорить о том, что его заканселили, а gracefulCtx - что он протух. В любом случае, это стоит покрыть тестами.